### PR TITLE
Destructuring assignment removes undefined from type when default value is given

### LINKED ISSUE #9341: Error with strictNullChecks when destructuring.
Given the following TypeScript:
```ts
const a: { x?: number; } = { };
let x = 0;
({ x = 1 } = a); // error: Type 'number | undefined' is not assignable to 'number'.
```

However, the destructuring assignment has a default value assigned. This should be treated the same way that strict null checking treats `VariableDeclarationList`, which does not have this error:

```ts
const a: { x?: number; } = { };
let { x = 1 } = a; 
let y: number = x; // ok, x has type 'number'
```

**TypeScript Version:** 
nightly

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #9341

### CODE REVIEW

**Comment by sandersn** [comment]
@mhegazy I think you are the best person to look at this while @ahejlsberg is gone.

**Comment by mhegazy** [comment]
Looks good to me. i would give @ahejlsberg some more time to check it as well.

**Comment by sandersn** [comment]
Sure. @ahejlsberg do you want to take a look?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12544,6 +12544,12 @@ namespace ts {
             if (exprOrAssignment.kind === SyntaxKind.ShorthandPropertyAssignment) {
                 const prop = <ShorthandPropertyAssignment>exprOrAssignment;
                 if (prop.objectAssignmentInitializer) {
+                    // In strict null checking mode, if a default value of a non-undefined type is specified, remove
+                    // undefined from the final type.
+                    if (strictNullChecks &&
+                        !(getCombinedTypeFlags(checkExpressionCached(prop.objectAssignmentInitializer)) & TypeFlags.Undefined)) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Use `checkExpression` here, not `checkExpressionCached`. The latter should only be used (a) when you know that contextual types will never have an effect on the expression and (b) when other checks of the same expression also use the caching version. Neither are true here.

**Comment by ahejlsberg** [comment]
:+1: Once you make the indicated change.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

**Pull request merged**